pkg/crd/markers/topology.go

```diff
@@ -110,4 +124,21 @@ func (m MapType) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
 	return nil
 }
 
-func (l MapType) ApplyFirst() {}
+func (m MapType) ApplyFirst() {}
+
+func (s StructType) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
+	if schema.Type != "object" {
+		return fmt.Errorf("must apply structType to an object")
+	}
+
+	if s != "atomic" && s != "granular" {
+		return fmt.Errorf(`StructType must be either "granular" or "atomic"`)
+	}
+
+	p := string(s)
+	schema.XMapType = &p
+
+	return nil
+}
+
+func (s StructType) ApplyFirst() {}
```

Comment:
why is this needed?

Reply:
Mainly driven by this: https://github.com/kubernetes-sigs/controller-tools/blob/4752ed2de7d2fc1b6b18398bf26cf2ce6b53cd94/pkg/crd/markers/doc.go#L27-L29. If I got it right, it suggests that "top-level" types (e.g. lists, as opposed to list items) can implement this method to ensure that their schemas will be applied before types that don't implement it.

The comment also says it's `discouraged` and `a hack` so if all this is not quite right, I'm happy to remove it. No tests fail if I remove `ApplyFirst` for StructType, for what it's worth 🤔

Reply:
I would only use it if there is an actual dependency between two types (like we do for ListType). But I don't think that's the case here.

Reply:
Removed in https://github.com/apelisse/controller-tools/pull/2/commits/5d66dadae91d4ac76bfc72472903f5ab8826dc87